main.go

```diff
@@ -1076,6 +1086,158 @@ func main() {
 		return Redirect(fmt.Sprintf("/users/%s/books/%s", user.Slug, vars["isbn"]))
 	}, loggedinMiddleware)
 
+	POST("/users/export/{user}", func(w Response, r Request) Output {
+		vars := VARS(r)
+
+		user, err := Q.UserBySlug(r.Context(), vars["user"])
+		if err != nil {
+			return NotFound
+		}
+
+		// export books data
+		books, err := Q.ExportBooksByUserId(r.Context(), user.ID);
+		if err != nil {
+			return InternalServerError(err)
+		}
+
+		booksFile, _ := json.MarshalIndent(books, "", "")
+		err = ioutil.WriteFile(BOOK_EXPORT_PATH, booksFile, 0644)
+		if err != nil {
+			return InternalServerError(err)
+		}
+		
+		buf := new(bytes.Buffer)
+		writer := zip.NewWriter(buf)
```

Comment:
zip.NewWeiter takes io.Writer to you can use `w` as it is an io.writer . that way you can skip writing the zip file to the memory first and stream the data from the structures to the zip file directly.